ExampleMod/ExampleWorld.cs

```diff
@@ -102,400 +104,437 @@ public override TagCompound Save()
+                traveller.active = false;
+                traveller.netSkip = -1;
+                traveller.life = 0;
+            }
+
+            // Main.time is set to 0 each morning, and only for one update. Sundialling will never skip past time 0 so this is the place for 'on new day' code
+            if (Main.dayTime && Main.time == 0)
+            {
+                // insert code here to change the spawn chance based on other conditions (say, npcs which have arrived, or milestones the player has passed)
+                // NPC won't spawn today if it stayed all night
+                if (traveller == null && Main.rand.Next(4) == 0) // 4 = 25% Chance
                 {
-                    // The traveler wont spawn at night so this is a safety if the chance is true but it somehow doesnt spawn
-                    randomTime = 0; // Set to 0 so the reset can happen
+                    // Here we can make it so the NPC doesnt spawn at the EXACT same time every time it does spawn
+                    travellerSpawnTime = GetRandomSpawnTime(5400, 8100); // minTime = 6:00am, maxTime = 7:30am
+
+                    // Set the items the NPC will spawn with. If the NPC respawns during the day, then it'll return with the same items.
+                    travelerItems = GetTravelerShop();
+                }
+                else
+                {
+                    travellerSpawnTime = double.MaxValue; // no spawn today
                 }
             }
-            else if (!Main.dayTime)
+
+            // Spawn the traveller after our chosen point in the day.
+            // We can't spawn hile the sundial is active, or while an event is running
+            // but if an event ends, the traveller can still arrive late in the day
+            bool eventActive = Main.eclipse || Main.invasionType > 0 && Main.invasionDelay == 0 && Main.invasionSize > 0;
+            if (traveller == null && Main.dayTime && Main.time > travellerSpawnTime && !Main.fastForwardTime && !eventActive)
+            {
+                int newExplorer = NPC.NewNPC(Main.spawnTileX * 16, Main.spawnTileY * 16, mod.NPCType("ExampleTravelingMerchant"), 1); // Spawning at the world spawn
+                traveller = Main.npc[newExplorer];
+                traveller.homeless = true;
+                traveller.direction = Main.spawnTileX >= WorldGen.bestX ? -1 : 1;
+                traveller.netUpdate = true;
+
+                // Annouce that the traveler has spawned in!
+                if (Main.netMode == NetmodeID.SinglePlayer) Main.NewText(Language.GetTextValue("Announcement.HasArrived", traveller.FullName), 50, 125, 255);
+                else NetMessage.BroadcastChatMessage(NetworkText.FromKey("Announcement.HasArrived", traveller.GetFullNetName()), new Color(50, 125, 255));
+            }
+        }
+
+        public static NPC FindNPC(int npcType)
+        {
+            foreach (NPC npc in Main.npc)

```

Comment:
I prefer changing it to a oneliner: `return Main.npc.FirstOrDefault(npc.type == npcType);`